File: fxgl/src/main/java/com/almasb/fxgl/scene/GameScene.java
/*
 * FXGL - JavaFX Game Library. The MIT License (MIT).
 * Copyright (c) AlmasB (almaslvl@gmail.com).
 * See LICENSE for details.
 */

package com.almasb.fxgl.scene;

import com.almasb.fxgl.app.GameApplication;
import com.almasb.fxgl.asset.FXGLAssets;
import com.almasb.fxgl.core.collection.Array;
import com.almasb.fxgl.core.collection.UnorderedArray;
import com.almasb.fxgl.core.logging.Logger;
import com.almasb.fxgl.effect.ParticleControl;
import com.almasb.fxgl.entity.*;
import com.almasb.fxgl.entity.component.ViewComponent;
import com.almasb.fxgl.entity.view.EntityView;
import com.almasb.fxgl.physics.PhysicsParticleControl;
import com.almasb.fxgl.ui.UI;
import javafx.collections.ObservableList;
import javafx.scene.Group;
import javafx.scene.Node;
import javafx.scene.canvas.Canvas;
import javafx.scene.canvas.GraphicsContext;
import javafx.scene.effect.BlendMode;
import javafx.scene.paint.Color;
import javafx.scene.text.Text;
import javafx.scene.transform.Rotate;
import javafx.scene.transform.Scale;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * Represents the scene that shows game objects on the screen during "play" mode.
 * Contains 3 layers. From bottom to top:
 * <ol>
 *     <li>Entities and their render layers (game view)</li>
 *     <li>Particles</li>
 *     <li>UI Overlay</li>
 * </ol>
 *
 * @author Almas Baimagambetov (AlmasB) (almaslvl@gmail.com)
 */
public final class GameScene extends FXGLScene
        implements EntityWorldListener, ModuleListener {

    private static final Logger log = Logger.get("FXGL.GameScene");

    /**
     * Root for entity views, it is affected by viewport movement.
     */
    private Group gameRoot = new Group();

    /**
     * Canvas for particles to accelerate drawing.
     */
    private Canvas particlesCanvas = new Canvas();

    /**
     * Graphics context for drawing particles.
     */
    private GraphicsContext particlesGC = particlesCanvas.getGraphicsContext2D();

    private Array<ParticleControl> particles = new UnorderedArray<>(16);

    /**
     * The overlay root above {@link #gameRoot}. Contains UI elements, native JavaFX nodes.
     * uiRoot isn't affected by viewport movement.
     */
    private Group uiRoot = new Group();

    private Text profilerText = new Text();

    public Text getProfilerText() {
        return profilerText;
    }

    protected GameScene(int width, int height) {
        getContentRoot().getChildren().addAll(gameRoot, particlesCanvas, uiRoot);

        initProfilerText(0, height - 120);
        initParticlesCanvas(width, height);
        initViewport(width, height);

        log.debug("Game scene initialized: " + width + "x" + height);
    }

    private void initProfilerText(double x, double y) {
        profilerText.setFont(FXGLAssets.UI_MONO_FONT.newFont(20));
        profilerText.setFill(Color.RED);
        profilerText.setTranslateX(x);
        profilerText.setTranslateY(y);

        uiRoot.getChildren().add(profilerText);
    }

    private void initParticlesCanvas(double w, double h) {
        particlesCanvas.setWidth(w);
        particlesCanvas.setHeight(h);
        particlesCanvas.setMouseTransparent(true);
    }

    private void initViewport(double w, double h) {
        Viewport viewport = getViewport();
        gameRoot.layoutXProperty().bind(viewport.xProperty().negate());
        gameRoot.layoutYProperty().bind(viewport.yProperty().negate());

        Scale scale = new Scale();
        scale.pivotXProperty().bind(viewport.xProperty());
        scale.pivotYProperty().bind(viewport.yProperty());
        scale.xProperty().bind(viewport.zoomProperty());
        scale.yProperty().bind(viewport.zoomProperty());
        gameRoot.getTransforms().add(scale);

        Rotate rotate = new Rotate(0, Rotate.Z_AXIS);
        rotate.pivotXProperty().bind(viewport.xProperty().add(w / 2));
        rotate.pivotYProperty().bind(viewport.yProperty().add(h / 2));
        rotate.angleProperty().bind(viewport.angleProperty().negate());
        gameRoot.getTransforms().add(rotate);
    }

//    /**
//     * Converts a point on screen to a point within game scene.
//     *
//     * @param screenPoint point in UI coordinates
//     * @return point in game coordinates
//     */
//    public Point2D screenToGame(Point2D screenPoint) {
//        return screenPoint
//                .multiply(1.0 / FXGL.getDisplay().getScaleRatio())
//                .add(getViewport().getOrigin());
//    }
//
//    /**
//     * Converts a point in game world to a point within screen (viewport).
//     *
//     * @param gamePoint point in game world coordinates
//     * @return point in UI coordinates
//     */
//    public Point2D gameToScreen(Point2D gamePoint) {
//        return gamePoint
//                .subtract(getViewport().getOrigin())
//                .multiply(FXGL.getDisplay().getScaleRatio());
//    }

    /**
     * @return unmodifiable list of UI nodes
     */
    public ObservableList<Node> getUINodes() {
        return uiRoot.getChildrenUnmodifiable();
    }

    /**
     * Add a node to the UI overlay.
     *
     * @param node UI node to add
     */
    public void addUINode(Node node) {
        uiRoot.getChildren().add(node);
    }

    /**
     * Add nodes to the UI overlay.
     *
     * @param nodes UI nodes to add
     */
    public void addUINodes(Node... nodes) {
        for (Node node : nodes)
            addUINode(node);
    }

    /**
     * Remove given node from the UI overlay.
     *
     * @param n node to remove
     * @return true iff the node has been removed
     */
    public boolean removeUINode(Node n) {
        return uiRoot.getChildren().remove(n);
    }

    /**
     * Remove nodes from the UI overlay.
     *
     * @param nodes nodes to remove
     */
    public void removeUINodes(Node... nodes) {
        for (Node node : nodes)
            removeUINode(node);
    }

    public void addUI(UI ui) {
        addUINode(ui.getRoot());
    }

    public void removeUI(UI ui) {
        removeUINode(ui.getRoot());
    }

    /**
     * Add a view to the game root.
     *
     * @param view view to add
     */
    public void addGameView(EntityView view, RenderLayer layer) {
        getRenderGroup(layer).getChildren().add(view);
    }

    public void addGameView(EntityView view) {
        getRenderGroup(RenderLayer.DEFAULT).getChildren().add(view);
    }

    /**
     * Remove a view from the game root.
     *
     * @param view view to remove
     */
    public void removeGameView(EntityView view, RenderLayer layer) {
        getRenderGroup(layer).getChildren().remove(view);
    }

    /**
     * Removes all nodes from the game view layer.
     */
    public void clearGameViews() {
        gameRoot.getChildren().clear();
    }

    /**
     * Removes all nodes from the UI overlay.
     */
    public void clearUINodes() {
        uiRoot.getChildren().clear();
    }

    /**
     * Set true if UI elements should forward mouse events
     * to the game layer.
     *
     * @param b flag
     * @defaultValue false
     */
    public void setUIMouseTransparent(boolean b) {
        uiRoot.setMouseTransparent(b);
    }

    /**
     * Returns graphics context of the game scene.
     * The render layer is over all entities.
     * Use this only if performance is required.
     * The drawing on this context can be done in {@link GameApplication#onUpdate(double)}.
     *
     * @return graphics context
     */
    public GraphicsContext getGraphicsContext() {
        return particlesGC;
    }

    /**
     * Returns render group for entity based on entity's
     * render layer. If no such group exists, a new group
     * will be created for that layer and placed
     * in the scene graph according to its layer index.
     *
     * @param layer render layer
     * @return render group
     */
    private Group getRenderGroup(RenderLayer layer) {
        Integer renderLayer = layer.index();

        Group group = null;

        for (Node n : gameRoot.getChildren()) {
            if ((int) n.getUserData() == renderLayer) {
                group = (Group) n;
                break;
            }
        }

        if (group == null) {
            log.debug("Creating render group for layer: " + layer.asString());

            group = new Group();
            group.setUserData(renderLayer);
            gameRoot.getChildren().add(group);
        }

        List<Node> tmpGroups = new ArrayList<>(gameRoot.getChildren());
        Collections.sort(tmpGroups, (g1, g2) -> (int) g1.getUserData() - (int) g2.getUserData());

        gameRoot.getChildren().setAll(tmpGroups);

        return group;
    }

    private boolean wasDirty = false;

    public void onUpdate(double tpf) {
        getViewport().onUpdate(tpf);

        boolean dirty = particles.isNotEmpty();

        if (dirty || wasDirty) {
            particlesGC.setGlobalAlpha(1);
            particlesGC.setGlobalBlendMode(BlendMode.SRC_OVER);

            // https://github.com/AlmasB/FXGL/issues/494
            particlesGC.clearRect(0, 0, getWidth(), getHeight());

            wasDirty = false;
        }

        for (ParticleControl particle : particles) {
            particle.renderParticles(particlesGC, getViewport().getOrigin());
        }

        wasDirty = dirty;
    }

    public void clear() {
        log.debug("Clearing game scene");

        getViewport().unbind();
        particles.clear();
        gameRoot.getChildren().clear();
        uiRoot.getChildren().clear();

        uiRoot.getChildren().add(profilerText);
    }

    @Override
    public void onEntityAdded(Entity entity) {
        initView(entity.getViewComponent());

        entity.addModuleListener(this);

        entity.getControlOptional(ParticleControl.class)
                .ifPresent(particles::add);
        entity.getControlOptional(PhysicsParticleControl.class)
                .ifPresent(particles::add);
    }

    @Override
    public void onEntityRemoved(Entity entity) {
        destroyView(entity.getViewComponent());

        entity.removeModuleListener(this);

        entity.getControlOptional(ParticleControl.class)
                .ifPresent(p -> particles.removeValueByIdentity(p));
        entity.getControlOptional(PhysicsParticleControl.class)
                .ifPresent(p -> particles.removeValueByIdentity(p));
    }

    private void initView(ViewComponent viewComponent) {
        EntityView view = viewComponent.getView();
        addGameView(view, viewComponent.getRenderLayer());

        viewComponent.renderLayerProperty().addListener((o, oldLayer, newLayer) -> {
            getRenderGroup(oldLayer).getChildren().remove(view);
            getRenderGroup(newLayer).getChildren().add(view);
        });
    }

    private void destroyView(ViewComponent viewComponent) {
        EntityView view = viewComponent.getView();
        removeGameView(view, viewComponent.getRenderLayer());
    }

    @Override
    public void onAdded(Control control) {
        if (control instanceof PhysicsParticleControl) {
            PhysicsParticleControl particleControl = (PhysicsParticleControl) control;
            particles.add(particleControl);
        }
    }

    @Override
    public void onRemoved(Control control) {
        if (control instanceof PhysicsParticleControl) {
            particles.removeValueByIdentity((PhysicsParticleControl) control);
        }
    }
}

File: fxgl/src/main/kotlin/com/almasb/fxgl/app/scene/GameScene.kt
/*
 * FXGL - JavaFX Game Library. The MIT License (MIT).
 * Copyright (c) AlmasB (almaslvl@gmail.com).
 * See LICENSE for details.
 */

package com.almasb.fxgl.app.scene

import com.almasb.fxgl.app.MainWindow
import com.almasb.fxgl.core.View
import com.almasb.fxgl.core.concurrent.Async
import com.almasb.fxgl.core.math.FXGLMath
import com.almasb.fxgl.dsl.FXGL.Companion.getAppHeight
import com.almasb.fxgl.dsl.FXGL.Companion.getAppWidth
import com.almasb.fxgl.dsl.FXGL.Companion.getSettings
import com.almasb.fxgl.entity.Entity
import com.almasb.fxgl.entity.EntityWorldListener
import com.almasb.fxgl.entity.GameWorld
import com.almasb.fxgl.entity.components.ViewComponent
import com.almasb.fxgl.logging.Logger
import com.almasb.fxgl.physics.PhysicsWorld
import com.almasb.fxgl.scene.Scene
import com.almasb.fxgl.ui.UI
import javafx.beans.property.SimpleIntegerProperty
import javafx.beans.value.ChangeListener
import javafx.collections.ObservableList
import javafx.event.EventHandler
import javafx.geometry.Point2D
import javafx.geometry.Rectangle2D
import javafx.scene.Group
import javafx.scene.Node
import javafx.scene.SceneAntialiasing
import javafx.scene.SubScene
import javafx.scene.input.MouseEvent
import javafx.scene.paint.Color
import javafx.scene.robot.Robot
import javafx.scene.shape.Rectangle
import javafx.scene.transform.Rotate
import javafx.scene.transform.Scale
import java.util.concurrent.Callable

/**
 * Represents the scene that shows entities on the screen during "play" mode.
 * Contains 2 layers. From bottom to top:
 *
 *  Entities and their render layers (game view)
 *  UI Overlay
 *
 *
 * @author Almas Baimagambetov (AlmasB) (almaslvl@gmail.com)
 */
class GameScene
@JvmOverloads
internal constructor(width: Int, height: Int,
                     val gameWorld: GameWorld,
                     val physicsWorld: PhysicsWorld,
                     private val is3D: Boolean = false) : FXGLScene(width, height), EntityWorldListener {

    companion object {
        private val log = Logger.get(GameScene::class.java)
    }

    /**
     * Root for entity views, it is affected by viewport movement.
     */
    private val gameRoot = Group()

    /**
     * The overlay root above [.gameRoot]. Contains UI elements, native JavaFX nodes.
     * uiRoot isn't affected by viewport movement.
     */
    private val uiRoot = Group()

    private val entities = ArrayList<Entity>()

    private var isZSortingNeeded = false

    private val zChangeListener = ChangeListener<Number> { _, _, _ ->
        isZSortingNeeded = true
    }

    /**
     * @return unmodifiable list of UI nodes
     */
    @get:JvmName("getUINodes")
    val uiNodes: ObservableList<Node>
        get() = uiRoot.childrenUnmodifiable

    private val updatableViews = arrayListOf<View>()

    val entitySelectionRectangle: EntitySelectionRectangle by lazy {
        EntitySelectionRectangle(this)
    }

    /**
     * If set to true, Game Scene will require calling step()
     * to advance each frame.
     */
    var isSingleStep = false

    val camera3D by lazy { Camera3D() }

    private val mouseWarper by lazy {
        Async.startAsyncFX(Callable { MouseWarper(window) }).await()
    }

    var isMouseGrabbed = false

    /**
     * If set to true, mouse movements will rotate the 3D camera as if it's an FPS camera.
     * Changing this value will also set [isMouseGrabbed] to the same value.
     */
    var isFPSCamera = false
        set(value) {
            field = value
            isMouseGrabbed = value

            if (value) {
                Async.startAsyncFX {
                    mouseWarper.warpToCenter()
                }
            }
        }

    private var lastMouseX = 0.0
    private var lastMouseY = 0.0

    init {
        contentRoot.children.addAll(
                if (is3D) make3DSubScene(width.toDouble(), height.toDouble()) else gameRoot,
                uiRoot
        )

        if (is3D) {
            val mouse3DHandler = EventHandler<MouseEvent> {
                if (isMouseGrabbed) {
                    // ignore warp mouse events
                    if (it.screenX.toInt() == mouseWarper.warpScreenX.toInt() && it.screenY.toInt() == mouseWarper.warpScreenY.toInt()) {
                        lastMouseX = it.screenX - window.x
                        lastMouseY = it.screenY - window.y

                        return@EventHandler
                    }
                }

                if (!isFPSCamera)
                    return@EventHandler

                val mouseX = it.screenX - window.x
                val mouseY = it.screenY - window.y

                val offsetX = mouseX - lastMouseX
                val offsetY = mouseY - lastMouseY

                // TODO: extract 100 and 0.5?
                if (FXGLMath.abs(offsetX) < 100 && FXGLMath.abs(offsetY) < 100) {
                    val mouseSensitivity = getSettings().mouseSensitivity

                    // only rotate if > 0.5 pixels
                    if (FXGLMath.abs(offsetX) > 0.5) {
                        if (mouseX > lastMouseX) {
                            camera3D.transform.lookRightBy(mouseSensitivity * (mouseX - lastMouseX))
                        } else if (mouseX < lastMouseX) {
                            camera3D.transform.lookLeftBy(mouseSensitivity * (lastMouseX - mouseX))
                        }
                    }

                    if (FXGLMath.abs(offsetY) > 0.5) {
                        if (mouseY > lastMouseY) {
                            val angle = mouseSensitivity * (mouseY - lastMouseY)

                            // use 85, rather than 90, to avoid potential rounding errors
                            if (camera3D.isOverRotationXAllowed || camera3D.transform.rotationX - angle > -85) {
                                camera3D.transform.lookDownBy(angle)
                            }

                        } else if (mouseY < lastMouseY) {
                            val angle = mouseSensitivity * (lastMouseY - mouseY)

                            if (camera3D.isOverRotationXAllowed || camera3D.transform.rotationX + angle < 85) {
                                camera3D.transform.lookUpBy(angle)
                            }
                        }
                    }
                }

                lastMouseX = mouseX
                lastMouseY = mouseY
            }

            input.addEventHandler(MouseEvent.MOUSE_MOVED, mouse3DHandler)
            input.addEventHandler(MouseEvent.MOUSE_DRAGGED, mouse3DHandler)
        }

        initViewport(width.toDouble(), height.toDouble())

        contentRoot.children.add(viewport.camera.viewComponent.parent)

        log.debug("Game scene initialized: " + width + "x" + height)

        gameWorld.addWorldListener(physicsWorld)
        gameWorld.addWorldListener(this)
    }

    private fun make3DSubScene(w: Double, h: Double): SubScene {
        val scene3D = SubScene(gameRoot, w, h, true, SceneAntialiasing.BALANCED)
        scene3D.camera = camera3D.perspectiveCamera

        return scene3D
    }

    private fun initViewport(w: Double, h: Double) {
        gameRoot.translateXProperty().bind(viewport.xProperty().negate())
        gameRoot.translateYProperty().bind(viewport.yProperty().negate())

        val scale = Scale()
        scale.pivotXProperty().bind(viewport.xProperty())
        scale.pivotYProperty().bind(viewport.yProperty())
        scale.xProperty().bind(viewport.zoomProperty())
        scale.yProperty().bind(viewport.zoomProperty())
        gameRoot.transforms.add(scale)

        val rotate = Rotate(0.0, Rotate.Z_AXIS)
        rotate.pivotXProperty().bind(viewport.xProperty().add(w / 2))
        rotate.pivotYProperty().bind(viewport.yProperty().add(h / 2))
        rotate.angleProperty().bind(viewport.angleProperty().negate())
        gameRoot.transforms.add(rotate)
    }

    override fun onUpdate(tpf: Double) {
        // if single step is configured, then step() will be called manually
        if (isSingleStep)
            return

        step(tpf)
    }

    fun step(tpf: Double) {
        gameWorld.onUpdate(tpf)
        physicsWorld.onUpdate(tpf)
        viewport.onUpdate(tpf)

        // update UI nodes
        updatableViews.forEach { it.onUpdate(tpf) }

        if (!is3D && isZSortingNeeded) {
            sortZ()
            isZSortingNeeded = false
        }

        if (is3D) {
            camera3D.update(tpf)
        }

        // TODO: extract 10?
        if (isMouseGrabbed && window.isFocused) {
            if (input.mouseXUI < 10) {
                mouseWarper.warpToCenter()
            } else if (input.mouseXUI + 10 > getAppWidth()) {
                mouseWarper.warpToCenter()
            } else if (input.mouseYUI < 10) {
                mouseWarper.warpToCenter()
            } else if (input.mouseYUI + 10 > getAppHeight()) {
                mouseWarper.warpToCenter()
            }
        }
    }

    /**
     * Add a node to the UI overlay.
     *
     * @param node UI node to add
     */
    fun addUINode(node: Node) {
        uiRoot.children.add(node)

        if (node is View)
            updatableViews += node
    }

    /**
     * Add nodes to the UI overlay.
     *
     * @param nodes UI nodes to add
     */
    fun addUINodes(vararg nodes: Node) {
        for (node in nodes)
            addUINode(node)
    }

    /**
     * Remove given node from the UI overlay.
     *
     * @param node node to remove
     */
    fun removeUINode(node: Node) {
        uiRoot.children.remove(node)

        if (node is View)
            updatableViews -= node
    }

    /**
     * Remove nodes from the UI overlay.
     *
     * @param nodes nodes to remove
     */
    fun removeUINodes(vararg nodes: Node) {
        for (node in nodes)
            removeUINode(node)
    }

    fun addUI(ui: UI) {
        addUINode(ui.root)
    }

    fun removeUI(ui: UI) {
        removeUINode(ui.root)
    }

    /**
     * Removes all nodes from the game view layer.
     */
    fun clearGameViews() {
        gameRoot.children.clear()
    }

    /**
     * Removes all nodes from the UI overlay.
     */
    fun clearUINodes() {
        uiRoot.children.clear()
        updatableViews.clear()
    }

    /**
     * Set true if UI elements should forward mouse events
     * to the game layer.
     *
     * @param b flag
     * @defaultValue false
     */
    fun setUIMouseTransparent(b: Boolean) {
        uiRoot.isMouseTransparent = b
    }

    private fun sortZ() {
        // it is important to sort in a different list since gameRoot is part of active scene graph
        // and does not allow duplicates that may occur during sorting
        val tmp = ArrayList(gameRoot.children)
        tmp.sortBy { (it.properties["viewData"] as GameView).z }

        gameRoot.children.setAll(tmp)
    }

    /**
     * Resets game world, physics world, game timer.
     * Unbinds viewport, clears game views and UI nodes.
     */
    internal fun reset() {
        log.debug("Clearing game scene")

        gameWorld.reset()

        // re-add listeners since above calls resets everything
        gameWorld.addWorldListener(physicsWorld)
        gameWorld.addWorldListener(this)

        physicsWorld.clear()
        physicsWorld.clearCollisionHandlers()

        timer.clear()

        entitySelectionRectangle.lastSelection.clear()

        viewport.unbind()
        gameRoot.children.clear()
        uiRoot.children.clear()
    }

    override fun onEntityAdded(entity: Entity) {
        entities.add(entity)
        initView(entity.viewComponent)
    }

    override fun onEntityRemoved(entity: Entity) {
        entities.remove(entity)
        destroyView(entity.viewComponent)
    }

    fun addGameView(view: GameView) {
        view.zProperty.addListener(zChangeListener)
        view.node.properties["viewData"] = view

        gameRoot.children.add(view.node)

        isZSortingNeeded = true
    }

    fun removeGameView(view: GameView) {
        view.node.properties.clear()

        gameRoot.children.remove(view.node)
    }

    private fun initView(viewComponent: ViewComponent) {
        val view = GameView(viewComponent.parent, viewComponent.zIndexProperty.value)
        view.zProperty.bind(viewComponent.zIndexProperty)

        addGameView(view)
    }

    private fun destroyView(viewComponent: ViewComponent) {
        // can we find a neater way to store node <-> z value pairings
        // since properties[] is null e.g. for particle views
        viewComponent.parent.properties["viewData"]?.let {
            val view = it as GameView
            view.zProperty.unbind()
            view.zProperty.removeListener(zChangeListener)

            removeGameView(view)
        }
    }

    private var isCursorInvisibleNeeded = false

    override fun onEnteredFrom(prevState: Scene) {
        if (isCursorInvisibleNeeded) {
            setCursorInvisible()
        }
    }

    override fun onExitingTo(nextState: Scene) {
        isCursorInvisibleNeeded = isCursorInvisible

        if (isCursorInvisibleNeeded) {
            val newCursor = preInvisibleCursor ?: window.defaultCursor

            newCursor?.let {
                setCursor(it)
            }
        }
    }

    /**
     * Must be constructed and used only on JavaFX App Thread.
     */
    private inner class MouseWarper(val window: MainWindow) {
        private val robot = Robot()

        var warpScreenX = 0.0
        var warpScreenY = 0.0

        /**
         * Warps the mouse cursor to window center.
         */
        fun warpToCenter() {
            warpScreenX = window.x + window.width / 2.0
            warpScreenY = window.y + window.height / 2.0

            robot.mouseMove(warpScreenX, warpScreenY)
        }
    }
}

class GameView(val node: Node, zIndex: Int) {
    val zProperty = SimpleIntegerProperty(zIndex)

    var z: Int
        get() = zProperty.value
        set(value) {
            zProperty.value = value
        }
}

class EntitySelectionRectangle(
    private val gameScene: GameScene
) : Rectangle(), View {

    private val world = gameScene.gameWorld
    private val input = gameScene.input

    /**
     * This list is populated on [stopSelection].
     */
    val lastSelection: MutableList<Entity> = arrayListOf()

    private var selectionStart: Point2D = Point2D.ZERO

    init {
        fill = Color.web("darkblue", 0.4)
        stroke = Color.LIGHTBLUE
        isVisible = false
    }

    fun startSelection() {
        if (scene == null) {
            gameScene.addUINode(this)
        }

        selectionStart = input.mousePositionUI
        translateX = selectionStart.x
        translateY = selectionStart.y
        width = 0.0
        height = 0.0

        isVisible = true
    }

    fun stopSelection() {
        isVisible = false

        lastSelection.clear()

        lastSelection.addAll(
            world.getEntitiesInRange(Rectangle2D(translateX, translateY, width, height))
        )
    }

    override fun onUpdate(tpf: Double) {
        if (isVisible) {
            val dx = input.mouseXUI - selectionStart.x
            val dy = input.mouseYUI - selectionStart.y

            if (dx > 0) {
                width = dx
            } else {
                translateX = input.mouseXUI
                width = -dx
            }

            if (dy > 0) {
                height = dy
            } else {
                translateY = input.mouseYUI
                height = -dy
            }
        }
    }

    override fun getNode(): Node {
        return this
    }

    override fun dispose() { }
}